cli/src/semgrep/formatter/sarif.py

```diff
@@ -14,12 +16,109 @@
 from semgrep.formatter.base import BaseFormatter
 from semgrep.rule import Rule
 from semgrep.rule_match import RuleMatch
+from semgrep.util import get_lines
 from semgrep.verbose_logging import getLogger
 
 logger = getLogger(__name__)
 
 
 class SarifFormatter(BaseFormatter):
+    @staticmethod
+    def _generate_taint_snippet(location: Any) -> Any:
+        content = "".join(
+            get_lines(Path(location.path), location.start.line, location.end.line + 2)
+        ).rstrip("\n")
+        snippet_before = "".join(
+            get_lines(
+                Path(location.path), location.start.line - 2, location.end.line - 1
+            )
+        ).lstrip("\n")
+        start_column = len(snippet_before) + location.start.col
+        end_column = len(snippet_before) + location.end.col
+        return "".join([snippet_before, content]), start_column, end_column
+
+    @staticmethod
+    def _taint_obj_intermediate_vars_to_thread_flow_location_sarif(
+        intermediate_var: Any, rule_match: RuleMatch
+    ) -> Any:
+        location = intermediate_var.location
+        content, start_column, end_column = SarifFormatter._generate_taint_snippet(
+            location
+        )
+        propagation_message_text = f"Propagator : '{content}' @ '{str(location.path)}:{str(location.start.line)}'"
+
+        intermediate_vars_location_sarif = {
+            "location": {
+                "message": {"text": propagation_message_text},
+                "physicalLocation": {
+                    "artifactLocation": {"uri": str(rule_match.path)},
+                    "region": {
+                        "startLine": location.start.line,
+                        "startColumn": start_column,
+                        "endLine": location.end.line,
+                        "endColumn": end_column,
+                        "snippet": {"text": content},
+                        "message": {"text": propagation_message_text},
+                    },
+                },
+            }
+        }
```

Comment:
Looks like duplicate code
Can you factorize?

Reply:
Done